src/bindings/c/ov/tests/ov_c_api_test.cpp

```diff
@@ -1648,71 +1646,120 @@ TEST(ov_model, ov_model_get_friendly_name) {
+    const char* str = "{1,?,300,40..100}";
+    ov_partial_shape_t* partial_shape = nullptr;
+
+    OV_ASSERT_OK(ov_partial_shape_init(&partial_shape, str));
+    auto tmp = ov_partial_shape_parse(partial_shape);
     EXPECT_STREQ(tmp, str);
 
-    delete tmp;
-    ov_partial_shape_free(&partial_shape);
+    ov_free(tmp);
+    ov_partial_shape_free(partial_shape);
 }
 
-TEST(ov_partial_shape, ov_partial_shape_init_and_parse_invalid) {
-    const char* str = "1,2+3;4..5";
+TEST(ov_partial_shape, ov_partial_shape_init_and_parse_dynamic_mix) {
+    const char* str = "{1,?,?,40..100}";
+    ov_partial_shape_t* partial_shape = nullptr;
+
+    OV_ASSERT_OK(ov_partial_shape_init(&partial_shape, str));
+    auto tmp = ov_partial_shape_parse(partial_shape);
+    EXPECT_STREQ(tmp, str);
+
+    ov_free(tmp);
+    ov_partial_shape_free(partial_shape);
+}
 
-    ov_partial_shape_t partial_shape;
-    partial_shape.ranks = 0;
+TEST(ov_partial_shape, ov_partial_shape_init_and_parse_dynamic_mix_2) {
+    const char* str = "{1,?,-1,40..100}";
+    ov_partial_shape_t* partial_shape = nullptr;
+    OV_ASSERT_OK(ov_partial_shape_init(&partial_shape, str));
+    auto tmp = ov_partial_shape_parse(partial_shape);
 
-    OV_EXPECT_NOT_OK(ov_partial_shape_init(&partial_shape,str));
+    ov_partial_shape_t* partial_shape2 = nullptr;
+    OV_ASSERT_OK(ov_partial_shape_init(&partial_shape2, tmp));
+    auto tmp2 = ov_partial_shape_parse(partial_shape);
+    EXPECT_STREQ(tmp, tmp2);
 
-    ov_partial_shape_free(&partial_shape);
+    ov_free(tmp);
+    ov_free(tmp2);
+    ov_partial_shape_free(partial_shape);
+    ov_partial_shape_free(partial_shape2);
 }
 
-TEST(ov_partial_shape, ov_partial_shape_to_shape) {
-    const char* str = "1,2,3,4,5";
+TEST(ov_partial_shape, ov_partial_shape_init_and_parse_dynamic_rank) {
+    const char* str = "?";
+    ov_partial_shape_t* partial_shape = nullptr;
 
-    ov_partial_shape_t partial_shape;
-    partial_shape.ranks = 0;
+    OV_ASSERT_OK(ov_partial_shape_init(&partial_shape, str));
+    auto tmp = ov_partial_shape_parse(partial_shape);
+    EXPECT_STREQ(tmp, str);
 
-    OV_ASSERT_OK(ov_partial_shape_init(&partial_shape,str));
+    ov_free(tmp);
+    ov_partial_shape_free(partial_shape);
+}
+
+TEST(ov_partial_shape, ov_partial_shape_init_and_parse_invalid) {
+    const char* str = "{1,2+3;4..5}";
+    ov_partial_shape_t* partial_shape = nullptr;
+
+    OV_EXPECT_NOT_OK(ov_partial_shape_init(&partial_shape, str));
+
+    ov_partial_shape_free(partial_shape);
+}
+
+TEST(ov_partial_shape, ov_partial_shape_to_shape) {
+    const char* str = "{10,20,30,40,50}";
+    ov_partial_shape_t* partial_shape = nullptr;
+
+    OV_ASSERT_OK(ov_partial_shape_init(&partial_shape, str));
 
     ov_shape_t shape;
     shape.ranks = 0;
-    OV_ASSERT_OK(ov_partial_shape_to_shape(&partial_shape, &shape));
+    OV_ASSERT_OK(ov_partial_shape_to_shape(partial_shape, &shape));
+    EXPECT_EQ(shape.ranks, 5);
```

Comment:
I haven't spotted it before but to align the naming convention, could you also rename the `ranks` member of the `ov_shape_t` structure and perhaps change it to `size_t`? The rank is between 0 and infinity so I think we don't need a signed type.
Alternatively you can change it the way `ov_partial_shape_t` is exposed (as a typedef) which would be even better in my opinion.